subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c

```diff
@@ -0,0 +1,612 @@
+#define DPS_TOPIC_OPERATION_ID_MAX_LEN		(DPS_OPERATION_ID_MAX_LEN + 100)
+#define DPS_REGISTERED_HUB_MAX_LEN		100
+#define DPS_SETTINGS_KEY			"azure_iot_hub"
+#define DPS_SETTINGS_HOSTNAME_LEN_KEY		"hostname_len"
+#define DPS_SETTINGS_HOSTNAME_KEY		"hostname"
+
+/* Periodically send messages to receive registration status.
+ * Keep polling as long as returned status code to $dps/registrations/res/#
+ * is 202. If it's 200, registration succeeded.
+ */
+#define DPS_TOPIC_REG_STATUS	\
+	"$dps/registrations/GET/iotdps-get-operationstatus/" \
+	"?$rid=%s&operationId=%s"
+#define DPS_TOPIC_REG		"$dps/registrations/res/"
+#define DPS_TOPIC_REG_SUB	DPS_TOPIC_REG "#"
+#define DPS_TOPIC_REG_PUB	"$dps/registrations/PUT/iotdps-register/" \
+				"?$rid=%s"
+
+#define DPS_REG_PAYLOAD "{registrationId:\"%s\"}"
+
+struct dps_reg_status {
+	enum dps_reg_state state;
+	struct k_delayed_work poll_work;
+	uint32_t retry;
+	/* Add 1 to make sure null termination can be used */
+	char reg_id[CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN + 1];
+	char operation_id[DPS_OPERATION_ID_MAX_LEN];
+	char assigned_hub[DPS_REGISTERED_HUB_MAX_LEN];
+	uint32_t assigned_hub_len;
+	struct k_sem registration_finished;
+	struct k_sem settings_loaded;
+};
+
+static dps_handler_t cb_handler;
+static struct dps_reg_status dps_reg_status;
+static struct mqtt_client *mqtt_client;
+static char dps_topic_reg_pub[sizeof(DPS_TOPIC_REG_PUB) +
+			      sizeof(dps_reg_status.reg_id)];
+
+/* Build time asserts */
+
+/* If DPS is enabled, the ID scope must be defined */
+BUILD_ASSERT(sizeof(CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE) - 1 > 0,
+	     "The DPS ID scope must be defined");
+
+/**@brief Callback when settings_load() is called. */
+static int dps_settings_handler(const char *key, size_t len,
+				settings_read_cb read_cb, void *cb_arg)
+{
+	int err;
+
+	if (strcmp(key, DPS_SETTINGS_HOSTNAME_LEN_KEY) == 0) {
+		err = read_cb(cb_arg, &dps_reg_status.assigned_hub_len,
+			      sizeof(dps_reg_status.assigned_hub_len));
+		if (err < 0) {
+			LOG_ERR("Failed to read hostname length, error: %d",
+				err);
+			return err;
+		}
+
+		return 0;
+	}
+
+	if (strcmp(key, DPS_SETTINGS_HOSTNAME_KEY) == 0) {
+		err = read_cb(cb_arg, dps_reg_status.assigned_hub,
+			      dps_reg_status.assigned_hub_len);
+		if (err < 0) {
+			LOG_ERR("Failed to read hostname, error: %d", err);
+			return err;
+		}
+
+		dps_reg_status.state = DPS_STATE_REGISTERED;
+
+		LOG_DBG("Azure IoT Hub hostname found, DPS not needed: %s",
+			log_strdup(dps_reg_status.assigned_hub));
+	}
+
+	return 0;
+}
+
+static int dps_save_hostname(const uint8_t *buf, uint32_t buf_len)
+{
+	int err;
+
+	err = settings_save_one(
+		DPS_SETTINGS_KEY "/" DPS_SETTINGS_HOSTNAME_LEN_KEY,
+		&buf_len, sizeof(buf_len));
+	if (err) {
+		return err;
+	}
+
+	err = settings_save_one(DPS_SETTINGS_KEY "/" DPS_SETTINGS_HOSTNAME_KEY,
+				buf, buf_len);
+	if (err) {
+		return err;
+	}
+
+	LOG_DBG("Azure IoT Hub acquired using DPS was successfully saved");
+
+	return 0;
```

Comment:
Could return err here. Consistent with ```dps_settings_init```

Reply:
Same reasoning as earlier, changed `dps_settings_init` to align